crates/exex/src/manager.rs

```diff
@@ -0,0 +1,466 @@
+use futures::StreamExt;
+use metrics::Gauge;
+use reth_metrics::{metrics::Counter, Metrics};
+use reth_primitives::BlockNumber;
+use reth_provider::CanonStateNotification;
+use reth_tracing::tracing::debug;
+use tokio::sync::{
+    mpsc::{self, error::SendError, Receiver, UnboundedReceiver, UnboundedSender},
+    watch,
+};
+use tokio_stream::wrappers::WatchStream;
+use tokio_util::sync::{PollSendError, PollSender};
+
+/// Metrics for an ExEx.
+#[derive(Metrics)]
+#[metrics(scope = "exex")]
+struct ExExMetrics {
+    /// The total number of canonical state notifications sent to an ExEx.
+    notifications_sent_total: Counter,
+    /// The total number of events an ExEx has sent to the manager.
+    events_sent_total: Counter,
+}
+
+/// A handle to an ExEx used by the [`ExExManager`] to communicate with ExEx's.
+///
+/// A handle should be created for each ExEx with a unique ID. The channels returned by
+/// [`ExExHandle::new`] should be given to the ExEx, while the handle itself should be given to the
+/// manager in [`ExExManager::new`].
+#[derive(Debug)]
+pub struct ExExHandle {
+    /// The execution extension's ID.
+    id: String,
+    /// Metrics for an ExEx.
+    metrics: ExExMetrics,
+
+    /// Channel to send [`CanonStateNotification`]s to the ExEx.
+    sender: PollSender<CanonStateNotification>,
+    /// Channel to receive [`ExExEvent`]s from the ExEx.
+    receiver: UnboundedReceiver<ExExEvent>,
+    /// The ID of the next notification to send to this ExEx.
+    next_notification_id: usize,
+
+    /// The finished block number of the ExEx.
+    ///
+    /// If this is `None`, the ExEx has not emitted a `FinishedHeight` event.
+    finished_height: Option<BlockNumber>,
+}
+
+impl ExExHandle {
+    /// Create a new handle for the given ExEx.
+    ///
+    /// Returns the handle, as well as a [`UnboundedSender`] for [`ExExEvent`]s and a
+    /// [`Receiver`] for [`CanonStateNotification`]s that should be given to the ExEx.
+    pub fn new(id: String) -> (Self, UnboundedSender<ExExEvent>, Receiver<CanonStateNotification>) {
+        let (canon_tx, canon_rx) = mpsc::channel(1);
+        let (event_tx, event_rx) = mpsc::unbounded_channel();
+
+        (
+            Self {
+                id: id.clone(),
+                metrics: ExExMetrics::new_with_labels(&[("exex", id)]),
+                sender: PollSender::new(canon_tx),
+                receiver: event_rx,
+                next_notification_id: 0,
+                finished_height: None,
+            },
+            event_tx,
+            canon_rx,
+        )
+    }
+
+    /// Reserves a slot in the `PollSender` channel and sends the notification if the slot was
+    /// successfully reserved.
+    ///
+    /// When the notification is sent, it is considered delivered.
+    fn send(
+        &mut self,
+        cx: &mut Context<'_>,
+        (event_id, notification): &(usize, CanonStateNotification),
+    ) -> Poll<Result<(), PollSendError<CanonStateNotification>>> {
```

Comment:
this type is basically a Sink,

we should also replicate the logic here with poll_reserve as a separate function the manager calls

Reply:
you want to implement sink?

Reply:
no, we should just add the equivalent fo poll_reserve,

but can also do this separately, so no blocker here